File: catherine_zhong/q2_FirstOccurrence.py
#time complexity: O(n)
#space complexity: O(n)

def main():
    test1 = "abracadabra"
    test2 = "Uber Career Prep"
    test3 = "zzyzx"
    test4 = "Billy Bobby Joe"

    print("test1: ", firstoccurrence(test1))
    print("test2: ", firstoccurrence(test2))
    print("test3: ", firstoccurrence(test3))
    print("test4: ", firstoccurrence(test4))

def firstoccurrence(word):
    unique_letters = set()
    result = []

    for char in word:
        if char not in unique_letters:
            unique_letters.add(char)
            result.append(char)
    
    return ''.join(result)

if __name__ == "__main__":
    main()

#time spent:  10 mins

File: homework0/nitai_mahat/q2_FirstOccurrence.java
package homework0.nitai_mahat;

import java.util.*;
/**
 * Time complexity = O(n) where n is the length of string
 * Space complexity = O(n) where n is the size of the hashset
 * 
 * 
 */
public class q2_FirstOccurrence {
    public static String firstOccurence(String word){
        StringBuilder finalString = new StringBuilder();
        Set<Character> uniqueSet = new HashSet<>();
    for(Character letter : word.toCharArray()){
        if(!uniqueSet.contains(letter)){
              finalString.append(letter);
        }
      
        uniqueSet.add(letter);
    }
    
   
    return finalString.toString();
    }
    //time spent: 8 min

    public static void main(String[] args){
        String test1 = "abracadabra";
        System.out.println(firstOccurence(test1));

        String test2 = "Uber CaPp";
        System.out.println(firstOccurence(test2));

        String test3 = "zzyzx";
        System.out.println(firstOccurence(test3));


    }
 
    
}
